File: 2024/python/day01/main.py
l, r = [], []  # noqa: E741

with open("input") as f:
    for line in f:
        x, y = map(int, line.strip().split())
        l.append(x)
        r.append(y)

l.sort()
r.sort()

# part1
total = 0
for a, b in zip(l, r):
    diff = abs(a - b)
    total += diff
print(total)

# part2
print(sum(x * r.count(x) for x in l))

File: 2024/r/day01/main.R
data <- read.delim("input", header = FALSE, sep = "", col.names = c("left", "right")) # nolint: line_length_linter.

data |>
  (\(df) {
    left <- sort(df$left)
    right <- sort(df$right)
    part1 <- sum(abs(left - right))
    message(part1)
  })()

data |>
  (\(df) {
    left_vals <- df$left
    right_counts <- table(df$right)
    part2 <- sum(sapply(left_vals, function(x) {
      count <- right_counts[as.character(x)]
      x * ifelse(is.na(count), 0, count)
    }))
    message(part2)
  })()
